### ISSUE TITLE
False positive "too few attributies for target variables" for altered table

### ISSUE THREAD

**Issue by zgrannan** [okbob/plpgsql_check#33]
Given the following SQL code:
```sql
CREATE TABLE my_table(r1 int, r2 int);
CREATE FUNCTION f() RETURNS INT
LANGUAGE PLPGSQL
AS $_$
DECLARE
record my_table;
BEGIN
  SELECT * INTO record FROM my_table;
  RETURN record.r1;
END
$_$;

ALTER table my_table drop column r2;
SELECT * FROM plpgsql_check_function('f()');
```
The following output is produced:
```
                         plpgsql_check_function                         
------------------------------------------------------------------------
 warning:00000:5:SQL statement:too few attributies for target variables
 Detail: There are more target variables than output columns in query.
 Hint: Check target variables in SELECT INTO statement.
```
I may be wrong about this, but I believe this warning is unnecessary?
It seems to me like the cause of the problem is the `ALTER TABLE` command.

Additionally, `attributies` should become `attributes`

**Comment by okbob** [comment]
Hi

2018-04-19 9:31 GMT+02:00 Zack Grannan <notifications@github.com>:

> Given the following SQL code:
>
> CREATE TABLE my_table(r1 int, r2 int);CREATE FUNCTION f() RETURNS INT
> LANGUAGE PLPGSQLAS $_$
> DECLARE
> record my_table;BEGIN
>   SELECT * INTO record FROM my_table;
>   RETURN record.r1;
> END
> $_$;
> ALTER table my_table drop column r2;SELECT * FROM plpgsql_check_function('f()')
>
> The following output is produced:
>
>                          plpgsql_check_function
> ------------------------------------------------------------------------
>  warning:00000:5:SQL statement:too few attributies for target variables
>  Detail: There are more target variables than output columns in query.
>  Hint: Check target variables in SELECT INTO statement.
>
> It seems to me like the cause of the problem is the ALTER TABLE command.
>
I am not able to reproduce it on master branch. Please, recheck it.


> Additionally, attributies should become attributes
>

fixed, thank you for report

**Comment by okbob** [comment]
2018-04-19 10:04 GMT+02:00 Pavel Stehule <pavel.stehule@gmail.com>:

> Hi
>
> 2018-04-19 9:31 GMT+02:00 Zack Grannan <notifications@github.com>:
>
>> Given the following SQL code:
>>
>> CREATE TABLE my_table(r1 int, r2 int);CREATE FUNCTION f() RETURNS INT
>> LANGUAGE PLPGSQLAS $_$
>> DECLARE
>> record my_table;BEGIN
>>   SELECT * INTO record FROM my_table;
>>   RETURN record.r1;
>> END
>> $_$;
>> ALTER table my_table drop column r2;SELECT * FROM plpgsql_check_function('f()')
>>
>> The following output is produced:
>>
>>                          plpgsql_check_function
>> ------------------------------------------------------------------------
>>  warning:00000:5:SQL statement:too few attributies for target variables
>>  Detail: There are more target variables than output columns in query.
>>  Hint: Check target variables in SELECT INTO statement.
>>
>> It seems to me like the cause of the problem is the ALTER TABLE command.
>>
> I am not able to reproduce it on master branch. Please, recheck it.
>

what Postgres version do you use?

**Comment by zgrannan** [comment]
I'm using 9.6.8, just tried again with `master` and the problem persists for me:

To build, I used the following Dockerfile:
```dockerfile
FROM library/postgres:9.6.8

ENV PLPGSQL_CHECK_VERSION 111c649a48706009e58b5ed16838cbbd2fe5c188

RUN buildDependencies="build-essential \
    ca-certificates \
    curl \
    git-core \
    postgresql-server-dev-$PG_MAJOR" \
  && apt-get update \
  && apt-get install -y --no-install-recommends ${buildDependencies} \
  && mkdir /tmp/build \
  && curl -o /tmp/build/check.tar.gz -SL \
  "https://github.com/okbob/plpgsql_check/archive/$PLPGSQL_CHECK_VERSION.tar.gz" \
  && cd /tmp/build && tar -xvzf check.tar.gz \
  && cd /tmp/build/plpgsql_check-$PLPGSQL_CHECK_VERSION \
  && make USE_PGXS=1 all \
  && make USE_PGXS=1 install \
  && rm -rf /tmp/build /var/lib/apt/lists/* 
```

**Comment by okbob** [comment]
2018-04-19 10:50 GMT+02:00 Zack Grannan <notifications@github.com>:

> I'm using 9.6.8, just tried again with master and the problem persists
> for me:
>
> To build, I used the following Dockerfile:
>
> FROM library/postgres:9.6.8
> ENV PLPGSQL_CHECK_VERSION 111c649a48706009e58b5ed16838cbbd2fe5c188
> RUN buildDependencies="build-essential \    ca-certificates \    curl \    git-core \    postgresql-server-dev-$PG_MAJOR" \
>   && apt-get update \
>   && apt-get install -y --no-install-recommends ${buildDependencies} \
>   && mkdir /tmp/build \
>   && curl -o /tmp/build/check.tar.gz -SL \
>   "https://github.com/okbob/plpgsql_check/archive/$PLPGSQL_CHECK_VERSION.tar.gz" \
>   && cd /tmp/build && tar -xvzf check.tar.gz \
>   && cd /tmp/build/plpgsql_check-$PLPGSQL_CHECK_VERSION \
>   && make USE_PGXS=1 all \
>   && make USE_PGXS=1 install \
>   && rm -rf /tmp/build /var/lib/apt/lists/*
>
>
I found it - and I hope I repair it. Please, check it now.

**Comment by zgrannan** [comment]
Just tried with the latest version, still doesn't work for me I'm afraid.

**Comment by okbob** [comment]
It is strange - I tested it on 10 with same behave like 9.6. Tomorrow I'll
test it on 9.6



2018-04-19 11:48 GMT+02:00 Zack Grannan <notifications@github.com>:

**Comment by okbob** [comment]
I did check master (last commit) on PostgreSQL 9.6.8 and it is works. What is your output?

**Comment by zgrannan** [comment]
Running the following script:

```sql
CREATE EXTENSION plpgsql_check;
CREATE TABLE my_table(r1 int, r2 int);
CREATE FUNCTION f() RETURNS INT
LANGUAGE PLPGSQL
AS $_$
DECLARE
record my_table;
BEGIN
  SELECT * INTO record FROM my_table;
  RETURN record.r1;
END
$_$;

ALTER table my_table drop column r2;
SELECT * FROM plpgsql_check_function('f()');
```

Returns:

```
~/plpgsql_check (master) psql -h localhost -U postgres < test2.sql 
CREATE EXTENSION
CREATE TABLE
CREATE FUNCTION
ALTER TABLE
                        plpgsql_check_function                         
-----------------------------------------------------------------------
 warning:00000:5:SQL statement:too few attributes for target variables
 Detail: There are more target variables than output columns in query.
 Hint: Check target variables in SELECT INTO statement.
(3 rows)
```

However I noticed something interesting. If I close the connection and then re-open it, the command correctly returns nothing:

```
~/plpgsql_check (master) psql -h localhost -U postgres
psql (9.6.6, server 9.6.8)
Type "help" for help.

postgres=# SELECT * FROM plpgsql_check_function('f()');
 plpgsql_check_function 
------------------------
(0 rows)

postgres=# \q
```

So actually, I think this is fine for my use case. Since when I'm running this the DDL commands will be run in a different connection anyways.

**Comment by okbob** [comment]
2018-04-20 10:12 GMT+02:00 Zack Grannan <notifications@github.com>:

> Running the following script:
>
> CREATE EXTENSION plpgsql_check;CREATE TABLE my_table(r1 int, r2 int);CREATE FUNCTION f() RETURNS INT
> LANGUAGE PLPGSQLAS $_$
> DECLARE
> record my_table;BEGIN
>   SELECT * INTO record FROM my_table;
>   RETURN record.r1;
> END
> $_$;
> ALTER table my_table drop column r2;SELECT * FROM plpgsql_check_function('f()');
>
> Returns:
>
> ~/plpgsql_check (master) psql -h localhost -U postgres < test2.sql
> CREATE EXTENSION
> CREATE TABLE
> CREATE FUNCTION
> ALTER TABLE
>                         plpgsql_check_function
> -----------------------------------------------------------------------
>  warning:00000:5:SQL statement:too few attributes for target variables
>  Detail: There are more target variables than output columns in query.
>  Hint: Check target variables in SELECT INTO statement.
> (3 rows)
>
> However I noticed something interesting. If I close the connection and
> then re-open it, the command correctly returns nothing:
>
> ~/plpgsql_check (master) psql -h localhost -U postgres
> psql (9.6.6, server 9.6.8)
> Type "help" for help.
>
> postgres=# SELECT * FROM plpgsql_check_function('f()');
>  plpgsql_check_function
> ------------------------
> (0 rows)
>
> postgres=# \q
>
>
looks like some PostgreSQL strange bug. When I run this script in regress
tests, it is ok, when I run it interactive, then it is fail, because
PostgreSQL still have old tuple description.

**Comment by okbob** [comment]
2018-04-20 11:22 GMT+02:00 Zack Grannan <notifications@github.com>:

> Well, in any event I think this is solved for my use case. Thanks for your
> help!
>

Thank you for report

**Comment by okbob** [comment]
Now, I understand to this issue, and unfortunately I cannot it fix - I cannot to enforce recompilation before checking. It should be 100% fixed in PostgreSQL 11.
